### PR TITLE
feat: Add the email confirmation step for the trial request - MEED-2928 - meeds-io/MIP#102  

### PR DESCRIPTION
**Pull request by MayTekayaa**
This change will Send the email confirmation code without sign-up for the free trial form.

### CODE REVIEW

File: deeds-dapp-webapp/src/main/webapp/vue-app/dapp/layout/components/FreeTrialForm.vue [hunk 1]
```diff
@@ -127,6 +127,7 @@
               extra-class="body-2"
               email-prepend-icon="fa-envelope fa-1x"
               code-prepend-icon="fa-lock fa-1x"
+              without-load-wallet-email="true"
```
Comment (boubaker) [inline deeds-dapp-webapp/src/main/webapp/vue-app/dapp/layout/components/FreeTrialForm.vue]:
```suggestion
              anonymous
```

File: deeds-dapp-webapp/src/main/webapp/vue-app/dapp/js/authorizationCodeService.js [hunk 2]
```diff
@@ -37,8 +37,8 @@ export function sendEmailConfirmation(email) {
   });
 }
 
-export function isCodeValid(code) {
-  return fetch(`${window.parentAppLocation}/api/authorization`, {
+export function isCodeValid(code, email) {
+  return fetch(`${window.parentAppLocation}/api/authorization?email=${email}`, {
```
Comment (boubaker) [inline deeds-dapp-webapp/src/main/webapp/vue-app/dapp/js/authorizationCodeService.js]: Never add email, password, username in clear text inside a GET HTTP Method or any other HTTP URL Method.

```suggestion
  return fetch(`${window.parentAppLocation}/api/authorization`, {
```

File: deeds-dapp-webapp/src/main/webapp/vue-app/dapp/common/components/EmailField.vue [hunk 3]
```diff
@@ -148,7 +152,7 @@ export default {
     code() {
       if (this.code && Number.isInteger(Number(this.code)) && Number.isFinite(Number(this.code)) && this.code.length === 6) {
         this.sendingCode = true;
-        this.$authorizationCodeService.isCodeValid(this.code)
+        this.$authorizationCodeService.isCodeValid(this.code, this.email)
```
Comment (boubaker) [inline deeds-dapp-webapp/src/main/webapp/vue-app/dapp/common/components/EmailField.vue:155]: Not needed, the email should be added to HttpSession.

```suggestion
        this.$authorizationCodeService.isCodeValid(this.code)
```

File: deeds-dapp-service/src/main/java/io/meeds/deeds/web/rest/AuthorizationCodeController.java [hunk 4]
```diff
@@ -49,12 +49,16 @@ public class AuthorizationCodeController {
   public void generateCode(Principal principal,
                            @RequestParam("email")
                            String email) {
-    if (principal == null || StringUtils.isBlank(principal.getName()) || StringUtils.isBlank(email)) {
+    if (principal == null && StringUtils.isBlank(email)) {
```
Comment (boubaker) [inline deeds-dapp-service/src/main/java/io/meeds/deeds/web/rest/AuthorizationCodeController.java]: why should we allow `StringUtils.isBlank(email)` when `principal == null` ?

File: deeds-dapp-service/src/main/java/io/meeds/deeds/web/rest/AuthorizationCodeController.java [hunk 5]
```diff
@@ -65,13 +69,19 @@ public void generateCode(Principal principal,
   @ResponseStatus(value = HttpStatus.NO_CONTENT)
   public void checkValidity(Principal principal,
                             @RequestHeader(name = CODE_VERIFICATION_HTTP_HEADER, required = true)
-                            int code) {
-    if (principal == null || StringUtils.isBlank(principal.getName())) {
+                            int code,
+                            @RequestParam(name = "email")
+                            String email) {
```
Comment (boubaker) [inline deeds-dapp-service/src/main/java/io/meeds/deeds/web/rest/AuthorizationCodeController.java]: email not needed, you can store it in current `HTTPSession`

```suggestion
                            int code) {
```

File: deeds-dapp-webapp/src/main/webapp/vue-app/dapp/common/components/EmailField.vue [hunk 6]
```diff
@@ -110,6 +110,10 @@ export default {
     autofocus: {
       type: Boolean,
       default: false,
+    },
+    withoutLoadWalletEmail: {
```
Comment (boubaker) [inline deeds-dapp-webapp/src/main/webapp/vue-app/dapp/common/components/EmailField.vue]:
```suggestion
    anonymous: {
```

File: deeds-dapp-service/src/main/java/io/meeds/deeds/web/rest/AuthorizationCodeController.java (hunk 4 above)
Comment (boubaker) [inline deeds-dapp-service/src/main/java/io/meeds/deeds/web/rest/AuthorizationCodeController.java]: Until we implement a `Captcha` for this in frontend, we will have to make sure that the same `client IP` doesn't invoke this multiple times. To determine the Cient IP, you can use this:

```java
String clientIp = request.getHeader("X-FORWARDED-FOR");  
if (clientIp == null) {  
    clientIp = request.getRemoteAddr();  
}
```

The `clientIp` has to be stored in a cache in Service layer and we will have to check if it generated a code more than 10 times per minute (the number of times has to be configurable using properties).

File: deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java [hunk 7]
```diff
@@ -56,10 +57,18 @@ public class AuthorizationCodeService {
   @Value("${meeds.authorizationCode.maxCodeVerification:20}")
   private int                            maxCodeVerification;
 
+  @Getter
+  @Value("${meeds.generationCode.limit:10}")
+  private int                            generationCodeLimit;
```
Comment (boubaker) [inline deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java:62]: Use `maxCodeSending`

```suggestion
```

File: deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java [hunk 8]
```diff
@@ -56,10 +57,18 @@ public class AuthorizationCodeService {
...
+
   private Duration                       maxCodeValidity;
 
   private Map<String, AuthorizationCode> authorizationCodes = new ConcurrentHashMap<>();
 
+  private Map<String, Integer> codeGenerationCount          = new ConcurrentHashMap<>();
+
+  private final Map<String, Long> codeGenerationStamp       = new ConcurrentHashMap<>();
+
```
Comment (boubaker) [inline deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java:71]:
```suggestion
  private Map<String, AuthorizationCode> authorizationCodesByIp = new ConcurrentHashMap<>();

```

File: deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java [hunk 9]
```diff
@@ -97,13 +106,32 @@ public void init() {
    *                                  reached or maximum code verification has
    *                                  been reached
    */
-  public void generateCode(String key, String email, Object data) throws IllegalAccessException {
+  public void generateCode(String key, String email, Object data, String clientIp) throws IllegalAccessException {
+    long currentTimestamp = System.currentTimeMillis() / 1000;
+    int count = 0;
     if (!hasValidDateCode(key)) {
       authorizationCodes.put(key, newAuthorizationCode());
     }
```
Comment (boubaker) [inline deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java:114]:
```suggestion
    if (!hasValidDateCode(key)) {
      authorizationCodes.put(key, newAuthorizationCode());
    }
    if (!hasValidDateCodeByIp(clientIp)) {
      authorizationCodesByIp.put(clientIp, authorizationCodes.get(key));
    }
```

File: deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java [hunk 10]
```diff
@@ -97,13 +106,32 @@ public void init() {
...
+            }
+        });
+    if (codeGenerationCount.containsKey(clientIp)) {
+      count = codeGenerationCount.get(clientIp);
+      if (count >= generationCodeLimit) {
+        throw new IllegalAccessException("Code generation limit exceeded");
+      }
+    } else {
+      codeGenerationStamp.put(clientIp,currentTimestamp);
+    }
```
Comment (boubaker) [inline deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java]: Externalize this in a dedicated method `checkClientIpBlackListed` + **throw the exception** with the ip address when blacklisted.

Comment (MayTekayaa) [reply deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java]: done!

**Review by boubaker — APPROVED** [review]

**Pull request merged**